services/auth/httpsign.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package auth
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/web/middleware"
+	"github.com/go-fed/httpsig"
+	"golang.org/x/crypto/ssh"
+)
+
+// Ensure the struct implements the interface.
+var (
+	_ Method = &HTTPSign{}
+	_ Named  = &HTTPSign{}
+)
+
+// HTTPSign implements the Auth interface and authenticates requests (API requests
+// only) by looking for http signature data in the "Signature" header.
+// more information can be found on https://github.com/go-fed/httpsig
+type HTTPSign struct {
+}
+
+// Name represents the name of auth method
+func (h *HTTPSign) Name() string {
+	return "httpsign"
+}
+
+// Verify extracts and validates HTTPsign from the Signature header of the request and returns
+// the corresponding user object on successful validation.
+// Returns nil if header is empty or validation fails.
+func (h *HTTPSign) Verify(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *models.User {
+	// HTTPSign authentication should only fire on API
+	if !middleware.IsAPIPath(req) {
+		return nil
+	}
+
+	sigHead := req.Header.Get("Signature")
+	if len(sigHead) == 0 {
+		return nil
+	}
+
+	if len(setting.SSH.TrustedUserCAKeys) == 0 {
+		return nil
+	}
+
+	validpk, err := VerifyCert(req)
+	if err != nil {
+		log.Error("VerifyCert failed:  %v", err)
```

Comment:
return err

Reply:
We likely additionally need to log with something like:

```go
log.Warn("Failed authentication attempt from %s")
```

with the ip address of the requester - so this can be added to the fail2ban configuration.

Reply:
> return err
can't reply err there, you probably mean `nil`

Reply:
> We likely additionally need to log with something like:
> 
> ```go
> log.Warn("Failed authentication attempt from %s")
> ```
> 
> with the ip address of the requester - so this can be added to the fail2ban configuration.

It's not done by any other authentication method though

Reply:
It is in UserSignin.

If you're providing a Signature you're trying to login with that signature if that signature fails that's a failed authentication attempt.

Reply:
ok, warning added